evap/grades/views.py

```diff
@@ -27,11 +27,17 @@ def index(request):
 
 
 def prefetch_data(courses):
-    courses = courses.prefetch_related("degrees", "responsibles")
+    courses = courses.prefetch_related("degrees", "responsibles", "evaluations", "grade_documents")
 
     course_data = []
     for course in courses:
-        course_data.append((course, course.midterm_grade_documents.count(), course.final_grade_documents.count()))
+        course_data.append(
+            (
+                course,
+                ilen(gd for gd in course.grade_documents.all() if gd.type == GradeDocument.Type.MIDTERM_GRADES),
+                ilen(gd for gd in course.grade_documents.all() if gd.type == GradeDocument.Type.FINAL_GRADES),
+            )
+        )
 
     return course_data
```

Comment:
Changed name, added type annotations, changed to use list comprehension.

Yes, we could just annotate the count directly in the database, but I'd reserve specialized query crafting for places where we really need it for performance. This page performs fine, so I'd just keep the code simple.
https://github.com/e-valuation/EvaP/blob/e889dce5b2111f42f3cda777649c19d30de165be/evap/staff/tools.py#L337

---
> In grades_semester_view, we include evaluation_badges, but that doesn't do anything really because no evaluation is set in the template environment 🤔

It adds an empty span if I'm not mistaken (`evaluation_badges.html` line 7). @janno42 is this a bug (missing content) or is this just leftover code?

Reply:
It's a bug, the course-related badges should be shown on this page a well. The badges probably need some refactoring to handle both courses and evaluations.